nni/algorithms/compression/pytorch/pruning/transformer_pruner.py

```diff
@@ -0,0 +1,306 @@
+        """
+        errmsg = 'Attention weight group sanity check not passed'
+        sparsity = None
+        for group in self.masking_groups:
+            assert len(group) == 4, errmsg + ': each group must have four weights'
+            assert group[0].module.weight.size() == group[1].module.weight.size() and \
+                group[1].module.weight.size() == group[2].module.weight.size(), \
+                errmsg + ': the dimensions of Q, K, V projection matrices must be the same '
+            assert group[0].module.weight.size()[0] == group[3].module.weight.size()[1], \
+                errmsg + ': the dimension of attention results must match with input for output projection'
+            assert group[0].config['sparsity'] == group[1].config['sparsity'] == \
+                   group[2].config['sparsity'] == group[3].config['sparsity'], \
+                errmsg + ': the sparsity of matrices in the same layer must be the same'
+            if sparsity is None:
+                sparsity = group[0].config['sparsity']
+            if self.global_sort:
+                assert sparsity == group[0].config['sparsity'], \
+                    errmsg + ': for global_sort=True, the sparsity for all modules must be the same'
+            t = group[0].module.weight.size(0) / self.head_hidden_dim
+            assert t % 1 == 0, errmsg + ': head_hidden_dim must be a divisor of the output dimension of the ' \
+                                        'projection weights'
+
+    def remove_ungrouped_modules(self):
+        """
+        Remove non-attention weights that might be mistakenly captured by a simplified config_list.
+        """
+        care_of_modules = set([x for layer in self.masking_groups for x in layer])
+        self.modules_wrapper = [x for x in self.modules_wrapper if x in care_of_modules]
+
+    def validate_config(self, model, config_list):
+        """
+        Parameters
+        ----------
+        model : torch.nn.Module
+            Model to be pruned
+        config_list : list
+            List on pruning configs
+        """
+        schema = CompressorSchema([{
+            'sparsity': And(float, lambda n: 0 < n < 1),
+            Optional('op_types'): [str],
+            Optional('op_names'): [str]
+        }], model, logger)
+
+        schema.validate(config_list)
+
+    def compress(self):
+        for pruning_iter in range(self.num_iterations):
+            if self.ranking_criterion in ['l1_activation', 'l2_activation', 'taylorfo']:
+                training = self.bound_model.training
+                self.bound_model.eval()
+                self._trainer(self.bound_model, optimizer=self._optimizer, criterion=self._criterion, epoch=0)
+                self.update_mask()
+                self.bound_model.train(training)
+            else:
+                self.update_mask()
+
+            # for iterative pruning, finetune before next iteration
+            if self.num_iterations > 1:
```

Comment:
maybe `if current_iteration < self.num_iterations:` is better?

Reply:
Then I think for one-shot pruning we also enter this condition because `0 < 1`?

Reply:
As we discussed yesterday, I think I will merge this condition with the condition below so that finetuning inside pruner will only be performed between iterations of iterative pruning, not after the last iteration, nor after one-shot pruning